### PR TITLE
fix(jellyfish-transaction): wrong encoding standard used on DCT_ID, AccountToUtxo.mintingOutputsStart and Vout.nTokenId

### LINKED ISSUE #1583: Wrong encoding standard on `DCT_ID`, `AccountToUtxo.mintingOutputsStart` and `Vout.nTokenId`
### What happened:

On `DeFiCh/ain` cpp node client; these 2 lines are different.

```cpp
READWRITE(COMPACTSIZE(v));
READWRITE(VARINT(v));
```

On jellyfish, we expected `VARINT` and `COMPACTSIZE` to be the commonly used [varuint-bitcoin](https://github.com/bitcoinjs/varuint-bitcoin). However, that's not true it's a different standard, `COMPACTSIZE != VARINT`.

When the value is `>128`, the variable is encoded differently.

VARINT: 129 = `0x8001`, MSB Base-128, https://en.wikipedia.org/wiki/LEB128
COMPACTSIZE: 129: `0x81`, Variable length integer, https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer

### Need to be fixed:

Going through the `DeFiCh/ain` cpp node client, these 3 implementations need to be swapped to the new standard. **Currently on MainNet, no transactions will cause it to fail since this issue is isolated to when the integer is `>128`, none of the integers are `>128` yet or will be anytime soon.**

- [x] Add VarInt support in `@defichain/jellyfish-buffer`: https://github.com/JellyfishSDK/jellyfish/pull/1586
- [ ] `DCT_ID`: https://github.com/JellyfishSDK/jellyfish/pull/1588
- [ ] `AccountToUtxo.mintingOutputsStart`: https://github.com/JellyfishSDK/jellyfish/pull/1588
- [ ] `Vout.nTokenId`: https://github.com/JellyfishSDK/jellyfish/pull/1588

/triage accepted
/priority urgent-now
/apps whale-api
/area apps packages
/assign @fuxingloh @ivan-zynesis @canonbrother

### PR DESCRIPTION
**Pull request by fuxingloh**
#### What this PR does / why we need it:

With the introduction of https://github.com/JellyfishSDK/jellyfish/pull/1586, this issue fixes the wrong use of `COMPACTSIZE` on `VARINT` implementation for `DCT_ID`, `AccountToUtxo.mintingOutputsStart` and `Vout.nTokenId`.

On jellyfish, we expected `VARINT` and `COMPACTSIZE` to be the commonly used [varuint-bitcoin](https://github.com/bitcoinjs/varuint-bitcoin). However, that's not true it's a different standard, `COMPACTSIZE != VARINT`. When the value is `>128`, the variable is encoded differently.

VARINT: 129 = `0x8001`, MSB Base-128
COMPACTSIZE: 129: `0x81`, commonly known as VarUInt on bitcoinjs

#### Which issue(s) does this PR fixes?:

Fixes #1583

### CODE REVIEW

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts [hunk 1]
```diff
@@ -3,17 +3,22 @@ import { BufferComposer, ComposableBuffer } from '@defichain/jellyfish-buffer'
 import { Script } from '../../tx'
 import { CScript } from '../../tx_composer'
 
-export interface TokenBalance {
+/**
+ * Known as "struct CBalances" in cpp.
+ */
+export interface TokenBalanceUInt32 {
```
Comment (ivan-zynesis) [inline packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:9]: Shall we keep the old interface naming? Seems possible for downstream  transition to new implementation without any regression.

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts [hunk 2]
```diff
@@ -3,17 +3,22 @@ import { BufferComposer, ComposableBuffer } from '@defichain/jellyfish-buffer'
...
   token: number // ---------------------| 4 bytes unsigned
```
Comment (ivan-zynesis) [inline packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:10]:
```suggestion
  token: number // ---------------------| VarInt{MSB-b128}
```
Theoretically MAX SAFE INTEGER in base128

**Review by ivan-zynesis — CHANGES_REQUESTED** [review]

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts (hunk 1 above)
Comment (fuxingloh) [reply packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:9]: How about we alias it, `TokenBalance = TokenBalanceUInt32`?

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts [hunk 3]
```diff
@@ -3,17 +3,22 @@ import { BufferComposer, ComposableBuffer } from '@defichain/jellyfish-buffer'
 import { Script } from '../../tx'
 import { CScript } from '../../tx_composer'
 
-export interface TokenBalance {
+/**
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:6]:
```suggestion
/**
 * @deprecated renamed to TokenBalanceUInt32
 */
export type TokenBalance = TokenBalanceUInt32

/**
```

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts (hunk 2 above)
Comment (fuxingloh) [reply packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:10]: Nope, there is another specification for "TokenBalanceUInt32" hence the rename. It's UInt32, hence 4 bytes unsigned.

File: packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts (hunk 1 above)
Comment (ivan-zynesis) [reply packages/jellyfish-transaction/src/script/dftx/dftx_balance.ts:9]: Yup that works. I just don't see necessity to break downstream for this.

File: packages/jellyfish-transaction/__tests__/index.test.ts [hunk 4]
```diff
@@ -46,3 +53,85 @@ it('should be able to use DeFiTransactionConstants constants to craft Transactio
   new CTransactionSegWit(data).toBuffer(buffer)
   expect(buffer.toBuffer().toString('hex')).toStrictEqual(hex)
 })
+
+it('should be able to parse a TestNet TxId: 68c7902823139c4f5e4e81993e366e3499c8c9a90b4e41dd1fc1290d3735b30d from height: 1077842', function () {
```
Comment (canonbrother) [inline packages/jellyfish-transaction/__tests__/index.test.ts:57]: yea just suggestion.. 
better adding the reference 
https://defiscan.live/transactions/68c7902823139c4f5e4e81993e366e3499c8c9a90b4e41dd1fc1290d3735b30d?network=TestNet
friendly for us to cross check the result

Comment (fuxingloh) [reply packages/jellyfish-transaction/__tests__/index.test.ts:57]: Can help me create a comment suggestions

**Review by canonbrother — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

**Review by ivan-zynesis — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Pull request merged**